ati_netcanoem_ft_driver/src/ati_netcanoem_ft_driver_node.ros2.cpp

```diff
@@ -0,0 +1,150 @@
+
+
+  const std::string reset_or_set_bias_service
+      = this->declare_parameter("reset_or_set_bias_service",
+                                DEFAULT_RESET_OR_SET_BIAS_SERVICE);
+
+  sensor_frame_
+      = this->declare_parameter("sensor_frame", DEFAULT_SENSOR_FRAME);
+
+  const uint8_t sensor_calibration_index
+      = static_cast<uint8_t>(this->declare_parameter(
+          "sensor_calibration_index", DEFAULT_SENSOR_CALIBRATION));
+
+  // Make the logging function
+  std::function<void(const std::string&)> logging_fn
+      = [this] (const std::string& message)
+  {
+    RCLCPP_INFO(this->get_logger(), "%s", message.c_str());
+  };
+  RCLCPP_INFO(
+      this->get_logger(),
+      "Connecting to ATI F/T sensor with CAN base ID %hhx on socketcan "
+      "interface %s...", sensor_base_can_id, can_interface.c_str());
+  sensor_ptr_
+      = std::unique_ptr<AtiNetCanOemInterface>(
+          new AtiNetCanOemInterface(logging_fn,
+                                    can_interface,
+                                    sensor_base_can_id));
+  const std::string serial_num = sensor_ptr_->ReadSerialNumber();
+  const auto firmware_version = sensor_ptr_->ReadFirmwareVersion();
+  RCLCPP_INFO(
+      this->get_logger(),
+      "Connected to sensor with serial # %s and firmware version %hhu "
+      "(major version) %hhu (minor version) %hu (build)",
+      serial_num.c_str(), firmware_version.MajorVersion(),
+      firmware_version.MinorVersion(), firmware_version.BuildNumber());
+  RCLCPP_INFO(
+      this->get_logger(),
+      "Attempting to load active calibration %hhu...",
+      sensor_calibration_index);
+  const bool set_calibration
+      = sensor_ptr_->LoadNewActiveCalibration(sensor_calibration_index);
+  if (set_calibration)
+  {
+    RCLCPP_INFO(
+        this->get_logger(),
+        "Loaded calibration %hhu",
+        sensor_calibration_index);
+    using geometry_msgs::msg::WrenchStamped;
+    status_pub_
+        = this->create_publisher<WrenchStamped>(status_topic, 1);
+    using std::placeholders::_1;
+    using std::placeholders::_2;
+    auto service_callback = std::bind(
+        &AtiNetCanOemDriverNode::ResetOrSetBiasCallback, this, _1, _2);
+    reset_or_set_bias_service_
+        = this->create_service<std_srvs::srv::SetBool>(
+            reset_or_set_bias_service, service_callback);
+    RCLCPP_INFO(
+        this->get_logger(),
+        "Starting to stream F/T measurements at %f Hz...", poll_rate);
+    poll_timer_
+        = rclcpp::create_timer(
+            this, this->get_clock(),
+            rclcpp::Duration::from_seconds(1. / poll_rate),
+            std::bind(&AtiNetCanOemDriverNode::PublishWrench, this));
+  }
+  else
+  {
+    RCLCPP_FATAL(
+        this->get_logger(),
+        "Failed to load calibration %x", sensor_calibration_index);
+  }
+}
+
+void AtiNetCanOemDriverNode::ResetOrSetBiasCallback(
+    std::shared_ptr<std_srvs::srv::SetBool::Request> req,
```

Comment:
As I understand, `req` can be a `const shared_ptr` here. Is a pair of `shared_ptr` the only signature available in ROS2 (e.g. nothing like `const RequestType& req, ResultType& res`)?

Reply:
@calderpg-tri see [here](https://github.com/ros2/rclcpp/blob/d3c0049b24f91b9bacdfa255dbdee8297862118c/rclcpp/include/rclcpp/any_service_callback.hpp#L202-L224) for supported service callback signatures. Not many options to choose from.